packages/shadydom/src/shadydom.js

```diff
@@ -74,7 +76,20 @@ if (utils.settings.inUse) {
     // Integration point with ShadyCSS to disable styling MutationObserver,
     // as ShadyDOM will now handle dynamic scoping.
     'handlesDynamicScoping': true,
-    'wrap': utils.settings.noPatch ? wrap : (n) => n,
+    // Ensure the node is wrapped. This should be used when `noPatch` is set
+    // to ensure Shadow DOM compatible DOM operation.
+    'wrap': utils.settings.noPatch ? wrap : patch,
+    // When code should be compatible with `noPatch` `true` and `on-demand`
+    // settings, `wrapIfNeeded` can be used for optimal performance (v. `wrap`)
+    // for all DOM operations except the following: `appendChild` and
+    // `insertBefore` (when the node is being moved from a location where it
+    // was logically positioned in the DOM); when setting `className`/`class`;
+    // when calling `querySelector|All`; when setting `textContent` or
+    // `innerHTML`; `addEventListener`; and all scope specific API's like
+    // `getRootNode`, `isConnected`, `slot`, `assignedSlot`, `assignedNodes`.
+    'wrapIfNeeded': utils.settings.noPatch === true ||
+      (utils.settings.noPatch === 'on-demand' && !utils.settings.hasDescriptors) ?
+      wrap : patch,
```

Comment:
Let's review the logic for what `wrap` and `wrapIfNeeded` get set to together (and cross-check against Polymer's versions)...